include/unix/osd.h

```diff
@@ -304,7 +306,46 @@ ofi_cpuid(unsigned func, unsigned subfunc, unsigned cpuinfo[4])
 	asm volatile("clflush %0" : "+m" (*(volatile char *) addr))
 #define ofi_sfence() asm volatile("sfence" ::: "memory")
 
-#else /* defined(__x86_64__) || defined(__amd64__) */
+#elif ( defined(__riscv) && defined(__riscv_xlen) && (__riscv_xlen == 64) && (defined(__linux__) || defined(__APPLE__)) )
+
+#include <stdio.h>
+#include <regex.h>
+
+static inline void
+ofi_cpuid(unsigned func, unsigned subfunc, unsigned cpuinfo[4])
+{
+     FILE *f = fopen("/proc/cpuinfo", "r");
+
+     regex_t r;
+     regmatch_t  match[1];
+
+     regcomp(&r, "hart", 0);
+     unsigned int hart = 0;
+     size_t lsz = 1024;
+     char *line = malloc(1024);
+     size_t rc = getline(&line, &lsz, f);
+     while(rc != -1 && 0 < rc) {
```

Comment:
nit: add space after while, if statements -- e.g. while (...)

Please use forward logic, i.e. rc > 0 to make it easier to read.